src/controllers/dynakube/version/reconciler.go

```diff
@@ -0,0 +1,126 @@
+package version
+
+import (
+	"context"
+	"fmt"
+	"os"
+	"path"
+	"time"
+
+	dynatracev1beta1 "github.com/Dynatrace/dynatrace-operator/src/api/v1beta1"
+	"github.com/Dynatrace/dynatrace-operator/src/dockerconfig"
+	"github.com/Dynatrace/dynatrace-operator/src/dtclient"
+	"github.com/Dynatrace/dynatrace-operator/src/kubeobjects"
+	"github.com/spf13/afero"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+const (
+	// ProbeThreshold is the minimum time to wait between version upgrades.
+	ProbeThreshold = 15 * time.Minute
+
+	TmpCAPath = "/tmp/dynatrace-operator"
+	TmpCAName = "dynatraceCustomCA.crt"
+)
+
+type Reconciler struct {
+	dynakube     *dynatracev1beta1.DynaKube
+	dtClient     dtclient.Client
+	hashFunc     ImageHashFunc
+	timeProvider *kubeobjects.TimeProvider
+
+	fs        afero.Afero
+	apiReader client.Reader
+}
+
+func NewReconciler(dynakube *dynatracev1beta1.DynaKube, apiReader client.Reader, dtClient dtclient.Client, fs afero.Afero, versionProvider ImageHashFunc, timeProvider *kubeobjects.TimeProvider) *Reconciler { //nolint:revive
+	return &Reconciler{
+		dynakube:     dynakube,
+		apiReader:    apiReader,
+		fs:           fs,
+		hashFunc:     versionProvider,
+		timeProvider: timeProvider,
+		dtClient:     dtClient,
+	}
+}
+
+// Reconcile updates the version and hash for the images used by the rec.Dynakube DynaKube instance.
+func (reconciler *Reconciler) Reconcile(ctx context.Context) error {
+	updaters := []versionStatusUpdater{
+		newActiveGateUpdater(reconciler.dynakube, reconciler.dtClient, reconciler.hashFunc),
+		newOneAgentUpdater(reconciler.dynakube, reconciler.dtClient, reconciler.hashFunc),
+		newCodeModulesUpdater(reconciler.dynakube, reconciler.dtClient),
+	}
+	if reconciler.needsReconcile(updaters) {
+		return reconciler.updateVersionStatuses(ctx, updaters)
+	}
+	return nil
+}
+
+func (reconciler *Reconciler) updateVersionStatuses(ctx context.Context, updaters []versionStatusUpdater) error {
+	dockerConfig, err := reconciler.createDockerConfigWithCustomCAs(ctx)
+	if err != nil {
+		return err
+	}
+
+	for _, updater := range updaters {
+		if !reconciler.needsUpdate(updater) {
+			continue
+		}
+		log.Info(fmt.Sprintf("updating %s version status", updater.Name()))
+		err := reconciler.run(updater, dockerConfig)
+		if err != nil {
+			log.Error(err, fmt.Sprintf("failed to update %s version status", updater.Name()))
+		}
+	}
+	return nil
+}
+
+func (reconciler *Reconciler) createDockerConfigWithCustomCAs(ctx context.Context) (*dockerconfig.DockerConfig, error) {
+	caCertPath := path.Join(TmpCAPath, TmpCAName)
+	dockerConfig := dockerconfig.NewDockerConfig(reconciler.apiReader, *reconciler.dynakube)
+	err := dockerConfig.SetupAuths(ctx)
+	if err != nil {
+		log.Info("failed to set up auths for image version checks")
+		return nil, err
+	}
+	if reconciler.dynakube.Spec.TrustedCAs != "" {
+		_ = os.MkdirAll(TmpCAPath, 0755)
```

Comment:
`os.MkdirTemp()`?

Reply:
this is not new code, it was just moved
and checking the docs for the `MkdirTemp` does not fit here, as it still has to be removed by us